Allow operator to ignore capacity

omarjackman opened the pull request:
This pull request adds an option for an operator to ignore their capacity and be assigned an additional chat.

omarjackman commented:
@beaucollins I'm having some issues with one of the tests here, It seems to be timing out. Can you help me figure out whats causing it?

beaucollins commented:
Looks like an unhandled rejection:
```
DEBUG=happychat:* npm test -- --grep system.info
```

And in the log I see:

```
  happychat:test:integration unhandled rejection { AssertionError: [ { online: true, ignoreCapacity: false, id: 'operator' } ] deepEqual [ { id: 'operator', online: true } ]
    at /Users/beaucollins/hc-service/test/unit/controller-test.js:266:6
    at EventEmitter.<anonymous> (/Users/beaucollins/hc-service/src/state/middlewares/socket-io/agents.js:28:3)
    at emitOne (events.js:96:13)
    at EventEmitter.emit (events.js:188:7)
    at EventEmitter.<anonymous> (/Users/beaucollins/hc-service/test/unit/controller-test.js:264:12)
    at emitOne (events.js:96:13)
    at EventEmitter.emit (events.js:188:7)
    at onAuthorized (/Users/beaucollins/hc-service/src/state/middlewares/socket-io/agents.js:41:9)
    at /Users/beaucollins/hc-service/src/state/middlewares/socket-io/agents.js:46:12
    at process._tickCallback (internal/process/next_tick.js:103:7)
  name: 'AssertionError',
  actual: [ { online: true, ignoreCapacity: false, id: 'operator' } ],
  expected: [ { id: 'operator', online: true } ],
  operator: 'deepEqual',
  message: '[ { online: true, ignoreCapacity: false, id: \'operator\' } ] deepEqual [ { id: \'operator\', online: true } ]',
  generatedMessage: true } +3ms
  happychat:test:mockio requesting room broadcast 0 +17ms
      1) should handle system.info event
```

Looks like it's missing the new `ignoreCapacity: false,` key in the assertion.

omarjackman commented:
@jordwest 

> Would it be possible to keep track of an "extra capacity" number or similar instead of a flag saying that the operator is ignoring capacity?

What would `extraCapacity` represent? Would it just be `load - capacity` where `load > capacity` for an operator?

> Without a capacity number, clicking +1 would leave the operator at the very end of the queue to receive chats

I'll look into this and get back to you

> how does the flag get reset?

It gets reset when the operator is assigned a chat in [src/state/middlewares/system/ignore-capacity-updater.js](https://github.com/Automattic/happychat-service/pull/121/commits/f5c9a333da188fa115f9e0fd06800db034846e63#diff-b27237c1169a2dbed8478aec367cd44e)

mattwondra commented:
So I could see us doing either @omarjackman's `ignoreCapacity` flag or @jordwest's `extraCapacity` number. But I think we do need to adjust [availability calculations](https://github.com/Automattic/happychat-service/blob/master/src/state/operator/selectors.js#L38-L39) for both scenarios (as @jordwest mentioned), because if `load` is over `capacity` as-is, then operators requesting an extra chat will never get it until all other operators are full.

#### Using `ignoreCapacity`

The biggest complexity with this approach is we need to do math to determine _how many_ above-capacity chats there are (since users can add as many extra chats as they want). If `load` exceeds `capacity` we can pretend that the current capacity _is_ `load`, so that when `ignoreCapacity` is flagged we add 1 to that capacity:

```js
const percentAvailable = ( { load, capacity, ignoreCapacity } ) => {
  let currentCapacity = ( load > capacity ) ? load : capacity;
  if ( ignoreCapacity ) {
    currentCapacity++;
  }
  return ( currentCapacity - defaultTo( 0, load ) ) / currentCapacity;
}

// Similar with totalAvailable()
```

This essentially sets up anyone above capacity as "100% booked" instead of "120%" or whatever, and then if you're "ignoring capacity" we push your fake capacity one higher.

#### Using `extraCapacity`

I imagine this would be a number showing how many extra chats the user can handle. The user state would look something like this:

```js
{
  ... user stuff ...,
  capacity: 3,
  extraCapacity: 1,
  load: 2
}
```

The [availability calculations](https://github.com/Automattic/happychat-service/blob/master/src/state/operator/selectors.js#L38-L39) would change to look like: 

```js
const percentAvailable = ( { load, capacity, extraCapacity } ) => ( capacity + extraCapacity - defaultTo( 0, load ) ) / ( capacity + extraCapacity )

// Similar with totalAvailable()
```

There _is_ some complexity in when you should decrement `extraCapacity` — If you decrement when a chat is assigned, it would be hard to get above 1 extra chat, e.g.:

```js
{ capacity: 3, extraCapacity: 0, load: 3 }
// HIT ME
{ capacity: 3, extraCapacity: 1, load: 3 }
// USER ASSIGNED, DECREMENT
{ capacity: 3, extraCapacity: 0, load: 4 }
// I'M BORED, HIT ME AGAIN
{ capacity: 3, extraCapacity: 1, load: 4 }
// UH OH I STILL DON'T HAVE EXTRA CAPACITY
```

Instead we'd probably want to decrement when a chat ends, if the load before was over `capacity`:

```js
{ capacity: 3, extraCapacity: 0, load: 3 }
// HIT ME
{ capacity: 3, extraCapacity: 1, load: 3 }
// ONE CHAT ENDS, KEEP EXTRA CAPACITY
{ capacity: 3, extraCapacity: 1, load: 2 }
// TWO CHATS START, KEEP EXTRA CAPACITY
{ capacity: 3, extraCapacity: 1, load: 4 }
// I'M BORED, HIT ME AGAIN
{ capacity: 3, extraCapacity: 2, load: 4 }
// ONE CHAT ENDS, SINCE load WAS ABOVE capacity DECREMENT extraCapacity
{ capacity: 3, extraCapacity: 1, load: 3 }
```

---

Ultimately I think both methods have merit. `ignoreCapacity` maps pretty cleanly to the HUD UI, where there's essentially a toggle you can enable/disable. `extraCapacity` feels a little more generalized and may come in handy in future unforeseen circumstances. I think I'd favor `extraCapacity` as it gives the state tree a little more transparency with what's actually going on IMO, but can't see major fault with either direction.

omarjackman commented:
@mattwondra I like these ideas but I'm worried we'd be digging our selves a nice hole if we add `extraCapacity`. It complicates things in my opinion to have to make a decision of which `capacity` property to decrement when a chat ends. Since a chat can end in more than one way, a timeout, `/end`. 

I think we can both agree that setting the `ignoreCapacity` flag to `true` at any time and then setting it to `false` when a chat is assigned is pretty simple. The complex part as you pointed out is computing availability. I feel like maintaining a capacity count adds one more point of failure when it comes to figuring out capacity and load. That means we need to update `capacityCount` when:
1. An op. increases/decreases their locale capacity
2. Op. goes active for a locale where they weren't active before.
3. Is assigned a chat
4. Ends a chat
5. Transfers a chat
6. And maybe more cases that I cannot think of.

`#1` and `#2` seems to be the most complex to handle unless I'm thinking about it incorrectly.

mattwondra commented:
@omarjackman Great points. You're definitely right, doing math when a chat is lost in one way or another does seem like a lot of added complexity. Is there more than one code path to a user being assigned a chat or does it always go through the same Redux action?

I'm :+1: with sticking with the boolean as a straightforward solution that doesn't try to optimize early for situations we might never run into. @jordwest — do you have strong opinions otherwise?

mattwondra commented:
@omarjackman Just found a bug — If I'm the only operator, with locale `en` checked but set to `0`, and I click the `+1` button in the HUD — in Calypso Happychat shows as available, but when I try to connect to an operator I get the "Sorry, we couldn't connect you with a Happiness Engineer" message and `chatStatus` is `pending`.

I don't think we ever expect an operator to be set at 0 in a locale they're "available" in, but this might be worth fixing.

mattwondra commented:
@omarjackman Sorry for the multiple threads but I'm noting things as I see them — 2 more oddities to report.

1. In the HUD when I'm RED or YELLOW, pushing `+1` to `ignoreCapacity` does the same thing as when I'm GREEN with 0 capacity — users see that there's availability, but when they try to connect the chat is marked "pending" and doesn't assign to my +1'd operator. I'd imagine we only want `ignoreCapacity` to work when you're GREEN.

2. One consequence of disabling `ignoreCapacity` on every chat assignment is that I can `+1` when I'm at 0 load, and as soon as one chat comes in the `+1` is disabled (even though I didn't "use" that +1). It complicates logic to only disable if a chat is assigned that puts you over capacity, but I'd think we might either want to do that, or solve in the HUD by only showing the `+1` button when you're at or above capacity.

jordwest commented:
I don't have a strong opinion for either approach. I think both `ignoreCapacity` and `extraCapacity` approaches have the issue that they don't really represent the operator's intention, but I'm not sure there's a better solution. For example: If the operator clicks the "+1" button, it only means they want one more chat than they _currently_ have, but we don't really want to make other assumptions (like assuming they're ignoring their capacity completely or changing some extra capacity setting).

In absence of a better solution though I think either approach is fine. We should make it clear in any case that it's a very temporary flag and could be reset at any moment, and possibly try to reset it as often as we can (its safer for the operator to have to regularly enable the flag than to accidentally leave the flag on and overload the operator)

jordwest commented:
Actually `ignoreCapacity` is growing on me, but I think naming it something different might make more sense, such as `requestingExtraChat`.

That way reading it in the code makes it clearer that it's the operator's intention to receive another chat.

For example if I was to see `requestingExtraChat` in the chat assignment prioritization, it would make more sense that that operator gets sent to the _front_ of the chat queue. If an operator is explicitly indicating they want another chat, they should be prioritised above other operators with availability.

As another example, when a new chat comes in it would make sense for `requestingExtraChat` to be reset; `ignoreCapacity` less so.

kriskarkoski commented:
I agree with both of Jordan's comments, we should err towards it being temporary and needing to use it frequently if you need to (otherwise you should just increase your throttle to always be at a higher number) and that the operator who sets it should go to the front of the queue for chat assignment since you would only use it when you feel you're in a good position with your existing chats and that takes a little pressure off the rest of the operators who may not be as comfortable to give them some breathing room before adding more chats.

omarjackman commented:
This pull request is ready for review again. I fixed the issues that @mattwondra brought up in https://github.com/Automattic/happychat-service/pull/121#issuecomment-294242539 and also changed the operator sorting algorithm so that operators who are requesting a chat (`requestingChat=true`) are the first to receive a pending chat.

omarjackman commented:
@jordwest, @beaucollins do you guys mind taking another look?

jordwest commented on src/state/operator/selectors.js:
```diff
@@ -82,14 +106,62 @@ const compareOperatorPriority = ( a, b ) => {
+		return statusComparison;
+	}
+
+	return compareOperatorsByChatLoad( a, b );
```
Another way we could write this function:

```js
const compareOperatorPriority = ( a, b ) => {
	// A function to convert a comparator to the calculated priority for the operators
	const mapComparators = map( f => f( a, b ) );
	// A function to find the first priority where both operators are not equal
	const firstPriority = find( p => p !== 0 );

	// List of priority comparators. The order of this list is important -
	// the first non-zero comparator will determine the priority for the
	// operators.
	const priorities = mapComparators( [
		compareByRequestForChat,
		compareOperatorByStatus,
		compareOperatorsByChatLoad,
	] );
	return firstPriority( priorities ) || 0;
}
```

What this will do is go through the list of comparators, and return the result of the first non-zero priority. So if `compareByRequestForChat` returns zero, it'll fall back to `compareOperatorByStatus` and so on.

This will let us clearly see and add/drop/reorder priority calculations in the future by simply changing the array elements.

This makes use of ramda's `find` and `map`, but could be done in lodash too (although I think ramda makes it a bit easier to read by making it possible to define a specialist `find`er and `map`per).

omarjackman replied:
addressed in a364577

jordwest commented on src/state/operator/selectors.js:
```diff
@@ -46,28 +45,53 @@ const totalAvailable = ( { load, capacity } ) => ( capacity - defaultTo( 0, load
+	}
+
+	// still have a tie so let "a" win
+	return -1;
```
If they're tied and the order doesn't actually matter, we can just return 0 and let `compareOperatorPriority` choose the priority. If we don't have useful information to prioritise them, it's fine to return 0 anyway and let the sorting algo choose the order.

jordwest commented on src/state/operator/selectors.js:
```diff
@@ -46,28 +45,53 @@ const totalAvailable = ( { load, capacity } ) => ( capacity - defaultTo( 0, load
+	const loadComparison = compareOperatorsByChatLoad( a, b );
+	if ( loadComparison !== 0 ) {
+		return loadComparison;
+	}
```
Instead of repeating the two above comparisons, we could just return 0 and let `compareOperatorPriority` do the other calcs. In fact I think this function could be reduced to just:

```js
const compareByRequestForChat = ( a, b ) => {
	if ( a.requestingChat !== b.requestingChat ) {
		return a.requestingChat ? -1 : 1;
	}

	return 0;
}
```

If neither or both of the operators are requesting an extra chat, then we don't have any useful information to determine their priority against each other.

omarjackman replied:
You are correct! I'd forgotten how the sort method was calling the compare function.

omarjackman replied:
addressed in b91425c780410e6879a0df7bb57ef4aa02360291

omarjackman commented:
> Some new tests in chatlist-assignment-test.js would also be really valuable

I added tests in faa92ef

> I think at this point it'd be worth moving those comparison funcs into their own file

Maybe we can do that in a separate clean up pull request.


@jordwest care to review again?

jordwest commented on src/state/middlewares/system/system-status-notifier.js:
```diff
@@ -5,17 +5,27 @@ import {
+	const requestingChat = hasOperatorRequestingChat( getState() );
+
+	const localeAvailabilityChanged = ! isEmpty( symmetricDifference( previous, current ) );
+	const requestingChatChanged = requestingChat === previouslyRequestingChat;
```
This should probably be `!==`, or the below action will be dispatched on every action

omarjackman replied:
This logic was replaced in 91872eb

jordwest commented on src/state/operator/selectors.js:
```diff
@@ -46,28 +48,33 @@ const totalAvailable = ( { load, capacity } ) => ( capacity - defaultTo( 0, load
-		return -1;
+const compareByRequestForChat = ( a, b ) => {
+	// using ! here incase one is undefined and the other is false
+	if ( ! a.requestingChat === ! b.requestingChat ) {
```
A common convention for casting to a boolean without flipping it is to use a double `!!`:

```js
if ( !! a.requestingChat === !! b.requestingChat ) {
```

omarjackman replied:
Correct but I made a conscious decision here to use the single `!` since the boolean value doesn't actually matter and this function is executed many times so I'm shaving two unnecessary ops from the stack

jordwest commented on src/state/operator/selectors.js:
```diff
@@ -82,14 +89,62 @@ const compareOperatorPriority = ( a, b ) => {
+
+	// get the first comparison thats truthy.
+	// Take advantage of -1 and 1 being truthy while 0 is falsy.
+	find( prioritizedComparisons, compare => result = compare( a, b ) );
```
Nice

jordwest commented:
Another option we could look into if perf is an issue is using memoization for selectors to avoid re-computation when the underlying state hasn't changed. Something like [reselect](https://github.com/reactjs/reselect) handles that automatically.

beaucollins commented:
@omarjackman master landed some conflicting changes. I merged master in and fixed the conflicts.

The biggest change touching this PR was in `system-status-notifier`. Instead of making the comparison on every single action dispatch it is now throttled to run at most once a second.

omarjackman commented:
> Instead of making the comparison on every single action dispatch it is now throttled to run at most once a second

Nice 👍 I was worried about that
